scantool/diag_general.c

```diff
@@ -323,52 +324,40 @@ diag_geterr(void) {
 
 /* Memory allocation */
 
-//diag_flcalloc (srcfilename, srcfileline, ptr, num,size) = allocate (num*size) bytes
-// also checks for size !=0
-// ret 0 if ok
-int diag_flcalloc(const char *name, const int line,
-	void **pp, size_t n, size_t s) {
-	void *p;
+// Stores pointer to a newly allocated buffer of n*s bytes to pp.
+// Also takes filename and line to report for debugging purposes.
+// Returns 0 in the absence of errors.
+int diag_fl_alloc(const char *fName, const int line,
+	void **pp, size_t n, size_t s, bool allocIsCalloc) {
+	char allocator;
+	char *errMsg = "%s:%d: %calloc(%ld, %ld) failed: %s\n";
 
-	//sanity check: make sure we weren't given a null ptr
-	//or a null size.
-
-	if ((s !=0) && (pp != NULL)) {
-		p = calloc(n, s);
+	if (allocIsCalloc) {
+		allocator = 'c';
 	} else {
-		p=NULL;
+		allocator = 'm';
 	}
 
-	*pp = p;
-
-	if (p == NULL) {
-		fprintf(stderr,
-			"%s:%d: calloc(%ld, %ld) failed: %s\n", name, line,
-			(long)n, (long)s, strerror(errno));
-		return diag_iseterr(DIAG_ERR_NOMEM);
+	// Check for null or overflow-causing parameters.
+	if (pp == NULL || n == 0 || s == 0 || (n*s)/s != n) {		// Is there a better way to check for overflow... ?
```

Comment:
checking for overflow : how about 
`if ((SIZE_MAX / s) < n) { overflow }`
saves a multiplication, and the intent is clear!

Reply:
You are right.

I checked GCC's and Clang's assembly output, hoping for them to be able to optimize both your and mine version of the overflow check to just a single multiplication and test of the overflow flag. Surprisingly, only GCC manages to do it, and only with your version. (This is with the -O3 or -Os flags, they give the same results in all cases.)